src/wp-includes/class-wp-scripts.php

```diff
@@ -714,6 +777,185 @@ public function in_default_dir( $src ) {
 		return false;
 	}
 
+	/**
+	 * This overrides the add_data method from WP_Dependencies, to support normalizing of $args.
+	 *
+	 * @param string $handle Name of the item. Should be unique.
+	 * @param string $key    The data key.
+	 * @param mixed  $value  The data value.
+	 * @return bool True on success, false on failure.
+	 */
+	public function add_data( $handle, $key, $value ) {
+		if ( 'script_args' === $key ) {
+			$args = $this->get_normalized_script_args( $handle, $value );
+			if ( $args['in_footer'] ) {
+				parent::add_data( $handle, 'group', 1 );
+			}
+			return parent::add_data( $handle, $key, $args );
+		}
+		return parent::add_data( $handle, $key, $value );
+	}
+
+	/**
+	 * Checks all handles for any delayed inline scripts.
+	 *
+	 * @return bool True if the inline script present, otherwise false.
+	 */
+	public function has_delayed_inline_script() {
+		foreach ( $this->registered as $handle => $script ) {
+			// non standalone after scripts of async or defer are usually delayed.
+			if ( in_array( $this->get_intended_strategy( $handle ), array( 'defer', 'async' ), true ) &&
+				$this->has_non_standalone_inline_script( $handle, 'after' )
+			) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Normalize the data inside the $args parameter and support backward compatibility.
+	 *
+	 * @param string        $handle Name of the script.
+	 * @param array         $args     {
+	 *      Optional. Additional script arguments. Default empty array.
+	 *
+	 *      @type boolean   $in_footer    Optional. Default true.
+	 *      @type string    $strategy     Optional. Values blocking|defer|async .Default 'blocking'.
+	 * }
+	 * @return array        Normalized $args array.
+	 */
+	private function get_normalized_script_args( $handle, $args = array() ) {
+		$default_args = array(
+			'in_footer' => false,
+			'strategy'  => 'blocking',
+		);
+		// Handle backward compatibility for $in_footer.
+		if ( true === $args ) {
+			$args = array( 'in_footer' => true );
+		}
+		return wp_parse_args( $args, $default_args );
+	}
+
+	/**
+	 * Get all of the scripts that depend on a script.
+	 *
+	 * @param string $handle The script handle.
+	 * @return array Array of script handles.
+	 */
+	private function get_dependents( $handle ) {
+		$dependents = array();
+
+		// Iterate over all registered scripts, finding ones that depend on the script.
+		foreach ( $this->registered as $registered_handle => $args ) {
+			if ( in_array( $handle, $args->deps, true ) ) {
+				$dependents[] = $registered_handle;
+			}
+		}
+		return $dependents;
+	}
+
+	/**
+	 * Get the strategy assigned during script registration.
+	 *
+	 * @param string $handle The script handle.
+	 * @return string|bool Strategy set during script registration. False if none was set.
+	 */
+	private function get_intended_strategy( $handle ) {
+		$script_args = $this->get_data( $handle, 'script_args' );
+		return isset( $script_args['strategy'] ) ? $script_args['strategy'] : false;
```

Comment:
If an unexpected strategy got registered somehow, I think it might be nice to throw a `_doing_it_wrong()` and make sure that we're only returning valid strategies. Here's a quick example, but could be improved.

```suggestion
		$strategy = isset( $script_args['strategy'] ) ? $script_args['strategy'] : false;
		
		// Ensure only valid strategies are returned.
		if ( ! in_array( $strategy, array( 'blocking', 'async', 'defer' ), true ) ) {
		 	// Optionally call _doing_it_wrong before returning.
		 	return false;
		}
		
		return $strategy;
```

Reply:
Maybe add a small helper function that returns the list of valid strategies (array( 'blocking', 'async', 'defer' )) so we can more cleanly extend this later.

Reply:
Addressing both in PR #56 (wip)

Reply:
Nice. I agree with Adam's suggestion on that file that we would need to add a filter so the list of valid strategies could be extended.

Reply:
Resolving to consolidate the conversation in #56.